### ISSUE TITLE
Unsound decimal behavior

### ISSUE THREAD

**Issue by toburger** [fable-compiler/Fable#1620]
### Description

I know decimals in fable aren't really decimals, but this was quite a surprise for me:

### Repro code

```fsharp

let items = [
    290.8M
    290.8M
    337.12M
    6.08M
   -924.8M
]

if List.sum items = 0M then
	// do something if sum is zero
else
	// do something else

```

[Link to Fable REPL](http://fable.io/repl2/#?code=LYewJgrgNgpgBAERgYwJbAIZQFDdgFzlXxmAGc4BeOAbWzgbgCYBOABgDoAOAWXsdade-BgGZRAdg4BGJn0ZwAbBzbCFAWhZMALNz4BdXAThkIwKnAAyqMvg6nzxUmVwAHAE6oAdvgBmXuAAiAFIAQUCTMzdPH38g4N8Ih1xUX0jzajYeOHwACxgvETgPbz8AwIciCgAvGHcQAEJA7BgoMhgiktjygHVQgBVm7CA&html=DwCwLgtgNgfAUHUBTAhgE3gAm54ElgqbhgAOAtEgI4CuAlgG4C8A5AMID2AdmEj+QBUAnqSQtMAY268erXgA8wAenDQA3JJAoATgGcCTGmABm5ABwssOYLona6pMJO0ddujvYDmdLpl3aJJgAiElJdAC4lJRouUgBrTwA6KQglbVQJMAABAEYANmiINDSMsETSFzQaTLpuRIgfRIArXSCYYCVbe0crbBs7Byc7V3cvHz8A4NCIqJj4pJSSlEzyNA4IXIKaIqWVtYhyyuqwWq56xpa2jq7B+A6QVAwEYAAjDjQhSSgUN2CUUgoUh4KB8SG0bTgOFwaEYmDoaGCSCgDV0IHI-1IQS+P10iORdFR6IBVyUMIYdyUbw+8EQKkgsDgQA&css=BYFwtgNgNAUARgewCYE8AEBvGadoGYIB2IAtAM4CWAXgKYBcaAjAGwAOAHgNwwC+MMAOjABDCoRIBjIiFGEaAJ0zZcSCmVYRhKBngg0uynAHcKSEMAaMADFYCk3XGmA0KAc1CWb9w2gBWAVzIQCjwUSWkaYgYJSJAFB1xhCDdxCjiwMmjY+J9dfRJVeRoJYKJohAh-MEJuPkExVn8QJUdWYSRVQlcGAQAmAFYisAScAmJyanomNgNHEzMLNAGrDhG0EXlXMRJEEBAEMEsh2pggA)

### Expected and actual results

I was expecting that the sum of the following values to be `0`.
The actual value is `1.1368683772161603e-13`.

This is quite unexpected for an F# developer where one of the language slogans is robustness.
The result is totally expected for floats, but it is a bug for decimals.

The following solutions would help avoid imprecise decimal values and gain trust in Fable:

- Disallow decimals. This sounds very radical, but since there is no equivalent replacement for decimals it is the best to give a compiler error before such kind of bugs can happen.
- Same as above, but give a warning.
- Replace decimals with something provided by Fable.

**Comment by alfonsogarciacaro** [comment]
We needed decimals to compile the F# compiler with Fable, compiling them as decimals was only a temporary solution but it's been working so we haven't "fixed" it yet. We do have a warning when explicitly converting floats to decimals but we don't for every use of decimals to prevent having too many decimals.

In any case I do agree we should implement decimals the correct way (as we do for int64), I wanted to do it for Fable 2 but other tasks were given higher priority at the end. The main part should be to port [this file](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Decimal.cs) to F#, any volunteers?

![](https://media.giphy.com/media/sXv0vaA4331Ti/giphy.gif)

**Comment by alfonsogarciacaro** [comment]
Or maybe using a tool to translate the C# code to JS? Or is there already a JS lib which is easy to integrate and has the proper semantics? Pinging @ncave for ideas.

**Comment by ncave** [comment]
@alfonsogarciacaro A lot of things can work, e.g. Bridge.NET uses [decimal.js](https://www.npmjs.com/package/decimal.js), or we can use F#'s [BigRational](https://github.com/fable-compiler/Fable/blob/master/src/js/fable-core/BigInt/q.fs), or port from C#, etc.

**Comment by ncave** [comment]
@alfonsogarciacaro Or, more appropriately, [decimal.js-light](https://github.com/MikeMcl/decimal.js-light/), which is the smaller version (without the trig. functions).

**Comment by toburger** [comment]
Hmm, maybe this is something I could contribute to Fable... 🤔 

Maybe we could port the implementation from elm: https://github.com/torreyatcitty/the-best-decimal/blob/1.0.0/src/Decimal.elm

This solution is based on BigInt (BigRational).
As BigInt isn't implemented in Fable this is something that has to be done before.
_An implementation of BigInt can be found here: https://github.com/cmditch/elm-bigint/blob/master/src/BigInt.elm_

What do you think, should we port elm's (functional) implementation or the C# (imperative?) code?

Slightly off topic:
There is at the moment an ongoing discussion at hackernews about how to deal with monetary values in JavaScript: https://frontstuff.io/how-to-handle-monetary-values-in-javascript and https://news.ycombinator.com/item?id=18334865
I think with a correct implementation of decimals we can do the right thing out of the box in Fable! 😄

**Comment by toburger** [comment]
#### Correction:

BigInt is implemented, but the literal doesn't seem to work:
* `bigint.Parse "12345678912345"` works and gives 12345678912345
* `12345678912345I` gives 1942903641

**Comment by ncave** [comment]
@toburger Thanks for the report, I opened an issue #1623 for it.

**Comment by alfonsogarciacaro** [comment]
Thanks for letting me know about BigRational, @ncave! This probably would be the easiest way to go as we already have BigInt (though with some problems in the REPL as @toburger points out) and is in line with our efforts to bring more F# modules to fable-core.

However, you're also right that this can have a performance hit (I guess especially in the REPL). Porting System.Decimal from C# is probably not an option as I just realized it uses unsafe code. So the main alternative would be to use decimal.js-light. I assume we would have to restructure the code as we did for long.js if we want to take advantage of tree shaking and have a better integration in the replacements module. It's a bit of work but I can do it (I'm only concerned about introducing bugs, we would need to port some tests from dotnet).

**Comment by toburger** [comment]
Fantastic work on the decimal support! Thank you! 👏 

I have a little quirk with `toLocaleString`, which doesn't work with decimals anymore.

```fsharp
let formatAsCurrency (d: decimal) =
    d?toLocaleString(locale, createObj [
        "style" ==> "currency"
        "currency" ==> "EUR"
    ])
```

This outputs (now) the plain value 3305.45 for instance...

The workaround is to converts it to a float before calling `toLocaleString`.

```fsharp
let formatAsCurrency (d: decimal) =
    (float d)?toLocaleString(locale, createObj [
        "style" ==> "currency"
        "currency" ==> "EUR"
    ])
```

which outputs the desired value 3.305,45 €.

**Comment by alfonsogarciacaro** [comment]
Wow, didn't know about the `toLocaleString` API, thanks for pointing it out! 😅 Well, it makes sense it doesn't work with decimals now. I'm reluctant to add support to System.Globalization in Fable because that'll be a lot of work, but it could be nice to add number formatting to [Fable.Powerpack](https://github.com/fable-compiler/Fable-powerpack) with a nice API. Would you like to PR it?

**Comment by toburger** [comment]
That would be really cool to have!

I've created some helper functions for my app but it would be really nice to have a nice statically typed API for formatting.

There exist four toLocaleString implementations for the following types:

* [Number.toLocaleString](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString)
* [Date.toLocaleString](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString)
* [Array.toLocalString](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/toLocaleString)
* [Object.toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toLocaleString)

I think it would make sense to implement them for numbers and dates.
Although I have no idea how a nice API could look like. Maybe something like this?

```fsharp
open Fable.PowerPack.Locale

Number.currency Euro // discriminated union?
|> Number.format 20.10 // use browsers locale

Number.percent
|> Number.withMinimumFractionDigits 2
|> Number.formatAs "de-DE" 0.10 // use specific locale

Date.date
|> Date.withMonthAs Long
|> Date.format (System.DateTime.Now)
```

**Comment by alfonsogarciacaro** [comment]
Looks good! @MangelMaxime added the date formatters to Fable.PowerPack, what do you think?
